db/models/user.js

```diff
@@ -25,18 +25,56 @@ const UserSchema = new Schema({
         validate: [({ length }) => length >= 6, "Password should be longer."]
     },
     designDiscipline: {
-        type: Schema.Types.ObjectId,
+        type: mongoose.Schema.Types.ObjectId,
         ref: "Disciplines"
     },
     authorisation: {
-        type: Schema.Types.ObjectId,
+        type: mongoose.Schema.Types.ObjectId,
         ref: "Authorisation"
     },
     project: [{
         type: Schema.Types.ObjectId,
         ref: "Project"
     }]
 });
+// use this function to hash the password before saving to database
+// this is mongoose midleware, once saved it will go to next
+// use old function wants to access to "this"
+UserSchema.pre('save',function(next){
+    // if password already modified(already hashed)
+    if(!this.isModified('password'))
+    return next();
+    // saltOrRound set up for 10 as doc example this mean how strong we need this encryption 
+    bcrypt.hash(this.password, 10, (err, passwordHash) => {
+        if(err)
+        return next(err);
+        this.password = passwordHash;
+        next();
+    });
+});
+
+//creating a comparePassword method as parameters password which is we are getting from the user(not hashed) and (cb as done) 
+UserSchema.methods.comparePassword = function(password, cb){
+    //  compare password, we getting from user and the one saved in database
+        bcrypt.compare(password, this.password, (err, isMatch) => {
+           // if we got any error with comparing
+            if(err)
+            //call cb with error
+            return cb(err);
+            // if there are no errors with compareing
+            else{
+                // if passwords match?
+                if(!isMatch)
+                // is not match call done function isMatch will be false
+                return cb(null, isMatch);
+                // if there is a match call done and return user object
+                // "this" mean user 
+                return cb(null, this);
```

Comment:
Will this return statement ever get executed? because it is after the return statement on line 69

Reply:
both return statement will get executed according to password is match or not, 69 will be under the if statement(if password isMatch false) and 72 will execute under the else statement(when password isMatch is true)

Reply:
Ok that makes sense